src/core/cluster.py

```diff
@@ -40,14 +63,33 @@ def client_relations(self) -> Set[Relation]:
     def unit_server(self) -> ZKServer:
         """The server state of the current running Unit."""
         return ZKServer(
-            relation=self.peer_relation, component=self.model.unit, substrate=self.substrate
+            relation=self.peer_relation,
+            data_interface=self.peer_unit_interface,
+            component=self.model.unit,
+            substrate=self.substrate,
         )
 
+    @property
+    def peer_units_data_interfaces(self) -> Dict[Unit, DataPeerOtherUnitData]:
+        """The cluster peer relation."""
+        if not self.peer_relation or not self.peer_relation.units:
+            return {}
+
+        for unit in self.peer_relation.units:
+            if unit not in self._servers_data:
+                self._servers_data[unit] = DataPeerOtherUnitData(
+                    model=self.model, unit=unit, relation_name=PEER
+                )
+        return self._servers_data
```

Comment:
**niptick, suggestion** what about using a cached_property instead of requiring an additional mutable attribute `_servers_data`? 

something like

```
    @cached_property
    def peer_units_data_interfaces(self) -> Dict[Unit, DataPeerOtherUnitData]:
        """The cluster peer relation."""
        if not self.peer_relation or not self.peer_relation.units:
            return {}
        return {
             unit: DataPeerOtherUnitData(model=self.model, unit=unit, relation_name=PEER)
             for unit in self.peer_relation.units
        }
           
```

Reply:
I support this comment, have implemented it verbatim and it works for Kafka, let's do that here too.

Reply:
Is it impossible to instatiate twice?
(On the 2nd attempt this will crash badly. We can only support a single instance of a `DataPeerOtherUnitData(model, unit, relation_name=PEER)` per unique parameters.)

Reply:
It's a cached property. Once referenced, it won't re-instantiate, unless done so in a different object. As there is only 1 object, that won't happen.

Reply:
Bah, I remember it now. Used long ago.

Sure, good call.

Reply:
Actually -- it'st not.

Note the differnce.

`cached_property` evaluates _once._. Meaning that additional units, after first evaluation will never be recognized.

My implementation maintains the data of known units (not to instantiate them once more as it would failr), while _also_ dynamically adds the new units that were not there before.